### PR TITLE
Java9

### PR DESCRIPTION
**Pull request by VoglSebastian**
EvoSuite is runnable with Java 9, still needs to be compiled with Java 8

### CODE REVIEW

File: client/src/main/java/org/evosuite/TestSuiteGenerator.java [hunk 1]
```diff
@@ -1,19 +1,19 @@
 /**
  * Copyright (C) 2010-2018 Gordon Fraser, Andrea Arcuri and EvoSuite
  * contributors
- *
+ * <p>
```
Comment (gofraser) [inline client/src/main/java/org/evosuite/TestSuiteGenerator.java]: Is this on purpose? There is a Maven plugin that updates all headers automatically, so if you want to change the copyright header we should use that and make a single commit for the copyright updates

File: client/src/main/java/org/evosuite/TestSuiteGenerator.java [hunk 2]
```diff
@@ -82,766 +87,766 @@
...
-					new GenericMethod(currentThreadMethod, currentThreadMethod.getDeclaringClass()), null,
-					Collections.emptyList());
-			VariableReference currentThreadVar = test.addStatement(currentThreadStmt);
-
-			Method getContextClassLoaderMethod = Thread.class.getMethod("getContextClassLoader");
-			Statement getContextClassLoaderStmt = new MethodStatement(test,
-					new GenericMethod(getContextClassLoaderMethod, getContextClassLoaderMethod.getDeclaringClass()),
-					currentThreadVar, Collections.emptyList());
-			VariableReference contextClassLoaderVar = test.addStatement(getContextClassLoaderStmt);
+    private static final String FOR_NAME = "forName";
```
Comment (gofraser) [inline client/src/main/java/org/evosuite/TestSuiteGenerator.java]: It's difficult to see what has changed since this change includes a reformatting. Is the reformatting on purpose, or can we reduce the change to the actual changes?

File: runtime/src/test/java/org/evosuite/runtime/util/JavaExecCmdUtilUnixTest.java [hunk 3]
```diff
@@ -41,6 +38,7 @@ public void unixNeverNull() {
     assertNotNull(JavaExecCmdUtil.getJavaBinExecutablePath(false));
   }
 
+  @Ignore
```
Comment (gofraser) [inline runtime/src/test/java/org/evosuite/runtime/util/JavaExecCmdUtilUnixTest.java]: Why?

File: runtime/src/test/java/org/evosuite/runtime/util/JavaExecCmdUtilUnixTest.java [hunk 4]
```diff
@@ -70,6 +68,7 @@ public void unixOldBehaviorJavaCmd() {
             JAVA_HOME_MOCK_PATH + SEPARATOR + "bin" + SEPARATOR + "java"));
   }
 
+  @Ignore
```
Comment (gofraser) [inline runtime/src/test/java/org/evosuite/runtime/util/JavaExecCmdUtilUnixTest.java]: Why?

File: runtime/src/test/java/org/evosuite/runtime/util/JavaExecCmdUtilUnixTest.java (hunk 3 above)
Comment (VoglSebastian) [reply runtime/src/test/java/org/evosuite/runtime/util/JavaExecCmdUtilUnixTest.java]: I thought this tests are faulty, because they fail if JAVA_HOME is not set.

Maybe we should test if JAVA_HOME is set first?

For Example:
assertNotNull("JAVA_HOME environment variable is not set.",JAVA_HOME_SYSTEM);

File: runtime/src/test/java/org/evosuite/runtime/util/JavaExecCmdUtilUnixTest.java (hunk 4 above)
Comment (VoglSebastian) [reply runtime/src/test/java/org/evosuite/runtime/util/JavaExecCmdUtilUnixTest.java]: Same as the other test

File: client/src/main/java/org/evosuite/TestSuiteGenerator.java (hunk 1 above)
Comment (VoglSebastian) [reply client/src/main/java/org/evosuite/TestSuiteGenerator.java]: The changes were not on purpose, i will remove this lines

File: client/src/main/java/org/evosuite/TestSuiteGenerator.java (hunk 2 above)
Comment (VoglSebastian) [reply client/src/main/java/org/evosuite/TestSuiteGenerator.java]: No, the reformatting was not on purpose, i will undo the reformatting and commit with only the actual changes.

File: runtime/src/test/java/org/evosuite/runtime/util/JavaExecCmdUtilUnixTest.java (hunk 3 above)
Comment (gofraser) [reply runtime/src/test/java/org/evosuite/runtime/util/JavaExecCmdUtilUnixTest.java]: Shouldn't this be handled by the @Before initTestEnvironment which sets JAVA_HOME?

Comment (VoglSebastian) [reply runtime/src/test/java/org/evosuite/runtime/util/JavaExecCmdUtilUnixTest.java]: line 18: 
private static final String JAVA_HOME_SYSTEM = System.getenv("JAVA_HOME");

This should be executed before initTestEnvironment().
Consequently JAVA_HOME_SYSTEM is null.

Leading to assertions in:
50:    assertFalse(StringUtils.isEmpty(JAVA_HOME_SYSTEM));
and
84-85: assertThat(JavaExecCmdUtil.getJavaBinExecutablePath(),
                  IsEqual.equalTo(JAVA_HOME_SYSTEM + SEPARATOR + "bin" + SEPARATOR + "java"));

Comment (gofraser) [reply runtime/src/test/java/org/evosuite/runtime/util/JavaExecCmdUtilUnixTest.java]: Indeed the JAVA_HOME environment variable may not be set and then this fails.

How about changing it to:

private static final String JAVA_HOME_SYSTEM = System.getProperty("java.home");

...or we could remove the assertions on JAVA_HOME_SYSTEM.

File: runtime/src/main/java/org/evosuite/runtime/util/ComputeClassWriter.java [hunk 5]
```diff
@@ -180,7 +184,10 @@ private boolean typeImplements(String type, ClassReader info, String itf)
      *             if the bytecode of 'type' cannot be found.
      */
     private ClassReader typeInfo(final String type) throws IOException, NullPointerException {
+        logger.info("Computing type info for: {}", type);
```
Comment (gofraser) [inline runtime/src/main/java/org/evosuite/runtime/util/ComputeClassWriter.java]: Remove debug info?

File: client/pom.xml [hunk 6]
```diff
@@ -147,6 +158,12 @@
             <artifactId>hibernate-core</artifactId>
             <scope>compile</scope>
         </dependency>
+        <dependency>
```
Comment (gofraser) [inline client/pom.xml]: Why is this needed?

File: master/pom.xml [hunk 7]
```diff
@@ -81,6 +96,27 @@
             <artifactId>logback-classic</artifactId>
             <optional>true</optional>
         </dependency>
+        <dependency>
```
Comment (gofraser) [inline master/pom.xml]: Are these additional dependencies on purpose?

File: runtime/pom.xml [hunk 8]
```diff
@@ -36,6 +56,11 @@
             <artifactId>ant-junit</artifactId>
             <scope>provided</scope>
         </dependency>
+        <dependency>
```
Comment (gofraser) [inline runtime/pom.xml]: Why do the dependencies change?

File: runtime/src/test/java/org/evosuite/runtime/util/JavaExecCmdUtilUnixTest.java (hunk 3 above)
Comment (VoglSebastian) [reply runtime/src/test/java/org/evosuite/runtime/util/JavaExecCmdUtilUnixTest.java]: Removing the assertion will only fix the issue for this test case.

But  JAVA_HOME_SYSTEM is also part of the only assertion of the unixNewBehavior() test. So removing this assertion, will lead to a test without assertions.

Comment (gofraser) [reply runtime/src/test/java/org/evosuite/runtime/util/JavaExecCmdUtilUnixTest.java]: So what about System.getProperty, which would make it work even if JAVA_HOME is not set?

(I am not sure I understand this test...)

Comment (VoglSebastian) [reply runtime/src/test/java/org/evosuite/runtime/util/JavaExecCmdUtilUnixTest.java]: I am not sure I understand the test or the class under test as well.

As far as i can say:
The method tested uses the JAVA_HOME environment variable if exists.
Otherwise it uses the current jvm path. (I dont know why)

So i would suggest to use JAVA_HOME if existing, otherwise java.home:

private static final String JAVA_HOME_SYSTEM = System.getenv("JAVA_HOME") != null ? System.getenv("JAVA_HOME") : System.getProperty("java.home");

Comment (gofraser) [reply runtime/src/test/java/org/evosuite/runtime/util/JavaExecCmdUtilUnixTest.java]: But if you set the JAVA_HOME environment variable, then the java.home property is initialised with that. So why not simply use java.home directly?

Comment (VoglSebastian) [reply runtime/src/test/java/org/evosuite/runtime/util/JavaExecCmdUtilUnixTest.java]: No,

System.getenv("JAVA_HOME")

always returns the value of the environment variable, while

System.getProperty("java.home")

returns the path to the JRE running the application.

See:
https://stackoverflow.com/questions/45441516/difference-between-java-home-and-java-home

Comment (gofraser) [reply runtime/src/test/java/org/evosuite/runtime/util/JavaExecCmdUtilUnixTest.java]: Fine, then let's go with your suggestion

File: runtime/pom.xml (hunk 8 above)
Comment (VoglSebastian) [reply runtime/pom.xml]: We need the SystemUtil class now in the ToolsJarLocator.

I have removed the other dependencies (as far as they are not needed)

File: runtime/src/main/java/org/evosuite/runtime/util/ComputeClassWriter.java (hunk 5 above)
Comment (VoglSebastian) [reply runtime/src/main/java/org/evosuite/runtime/util/ComputeClassWriter.java]: I have removed it

File: client/pom.xml (hunk 6 above)
Comment (VoglSebastian) [reply client/pom.xml]: It was not. I have removed it

File: master/pom.xml (hunk 7 above)
Comment (VoglSebastian) [reply master/pom.xml]: No, i have removed them

File: client/pom.xml [hunk 9]
```diff
@@ -147,6 +158,12 @@
             <artifactId>hibernate-core</artifactId>
             <scope>compile</scope>
         </dependency>
+        <!--<dependency>
```
Comment (gofraser) [inline client/pom.xml]: Sorry for nitpicking, but if this is commented out, why add it in the first place?

Comment (VoglSebastian) [reply client/pom.xml]: You are right, commented it out to test, if we need it to compile and forgot to remove it.
